File: massPayment.js
var fs = require('fs');
var request = require('request');

/**
 *  Put your settings here:
 *      - filename: file to which the payments for the mass payment tool are written
 *      - node: address of your node in the form http://<ip>:<port>
 *      - apiKey: the API key of the node that is used for distribution
 *      - feeAssetId: id of the asset used to pay the fee, null for Waves
 *      - fee: amount of fee to spend for the tx
 */
var config = {
    filename: '<your payments file>',
    node: 'http://<ip>:<port>',
    apiKey: 'put the apiKey for the node here',
    feeAssetId: null,
    fee: 10000
};

/**
 * The method that starts the payment process.
 */
var start = function() {
    var paymentsString = fs.readFileSync(config.filename).toString();
    var payments = JSON.parse(paymentsString);

    // start with the first payment, in case there is an error in the process,
    // it could just be restarted by changing the 0 to the the last successful
    // transmission increased by 1.
    doPayment(payments, 0);
};

/**
 * This method executes the actual payment transactions. One per second, so that the network
 * is not flooded. This could potentially be modified once the transaction limit of 100 tx
 * per block is raised.
 *
 * @param payments the array of payments (necessary to start this method recursively)
 * @param counter the current payment that should be done
 */
var doPayment = function(payments, counter) {
    var payment = payments[counter];

    if (config.feeAssetId !== null) {
        payment.feeAssetId = config.feeAssetId;
        payment.fee = config.fee;
    }

    setTimeout(function() {
        request.post({ url: config.node + '/assets/transfer', json: payment, headers: { "Accept": "application/json", "Content-Type": "application/json", "api_key": config.apiKey } }, function(err, response, body) {
            if (err) {
                console.log(err);
            } else {
                if (body.error) {
                    console.log('error during transfer: ' + body.message);
                } else {
                    console.log(counter + ' send ' + payment.amount + ' of ' + payment.assetId + ' to ' + payment.recipient + '!');
                }
                counter++;
                if (counter < payments.length) {
                    doPayment(payments, counter);
                }
            }
        });
    }, 1000);
};

start();

File: massPayment.py
import json
import time
import pywaves as pw
import time

'''
    Configuration section:
        privateKey: the private key of the address you want to distribute from
        file: the calculated payout file
        timeout: timeout between requests send to nodes in ms
        assetId: the id of the asset you want to distribute, '' for Waves
        nodes: a list of nodes to which the signed transactions should be send to, in the format: http://host:port
'''
config = {
	'privateKey': '',
	'file': '',
	# timeout between requests in ms
	'timeout': 20,
	'assetId': '',
	'nodes': [
	]
}

startTime = time.time()
totalPayments = 0
paid = 0
counter = 0

pw.setNode(config['nodes'][1], 'mainnet')
print(config['nodes'][0])
address = pw.Address(privateKey = config['privateKey'])

'''
    Method that actually generates the signed transactions and sends them to
    the configured nodes.
'''
def pay(batch):
	global paid
	global address
	global counter

	node = config['nodes'][counter % len(config['nodes'])]
	counter += 1
	pw.setNode(node, 'mainnet')

	print('number of payouts in batch: ' + str(len(batch)))
	print('batch: ' + str(batch))
	print('paid from address: ' + address.address)
	print('paid via node: ' + node)
	paid += len(batch)
	if (config['assetId'] != ''):
		print('paying in asset: ' + config['assetId'])
		tx = address.massTransferAssets(batch, pw.Asset(config['assetId']))
	else:
		print('paying in Waves!')
		tx = address.massTransferWaves(batch)
	print('tx: ' + str(tx))

with open(config['file']) as json_data:
	payments = json.load(json_data)
	currentBatch = []

	totalPayments = len(payments)
	for payment in payments:
		if (len(currentBatch) < 100):
			if (('assetId' in payment and payment['assetId'] != config['assetId']) or ('assetId' not in payment and config['assetId'] != '')):
				print('transaction with wrong assetId found!')
				exit()
			currentBatch.append({ 'recipient': payment['recipient'], 'amount': payment['amount'] })
		if (len(currentBatch) == 100):
			pay(currentBatch)
			currentBatch = []
			time.sleep(config['timeout'] / 1000)
	pay(currentBatch)
	print('planned payouts: ' + str(totalPayments) + ', paid: ' + str(paid))
	usedTime = time.time() - startTime
	print('time: ' + str(usedTime) + ' (' + str(totalPayments / usedTime) + 'transfers/s)')
